### ISSUE TITLE
Show analysis of alternate moves when replaying games

### ISSUE THREAD

**Issue by MarkZH** [MarkZH/Genetic_Chess#27]
Show scores and alternate play lines when replaying games by loading the Genetic_AI id.

Also, allow loading game by Round number for files with multiple games.

**Comment by MarkZH** [comment]
Specifically, what is up with this game?

```
[White "Genetic AI 10"]
[Black "Genetic AI 51"]
[Round "30"]
[Result "0-1"]

1. e3 d5 { d5 d4 }
2. Bb5+ Qd7 { Qd7 a4 }
3. g4 e5 { e5 d4 }
4. Bxd7+ Kxd7 { Kxd7 d4 }
5. b4 Bxb4 { Bxb4 f4 }
6. c4 dxc4 { dxc4 f4 }
7. Qa4+ Kd6 { Kd6 Qxa7 }
8. Qxb4+ c5 { c5 Qxc4 }
9. Qxb7 Bxg4 { Bxg4 Qxa7 }
10. d3 cxd3 { cxd3 Qxa7 }
11. Nc3 a5 { a5 Qxf7 }
12. Qd7+ Bxd7 { Bxd7 a3 }
13. e4 h5 { h5 a4 }
14. Bg5 c4 { c4 f4 }
15. Nb5+ Kc5 { Kc5 f4 }
16. Rd1 f5 { f5 Rxd3 }
17. Bf6 fxe4 { fxe4 Rxd3 }
18. h3 Bxh3 { Bxh3 Rxd3 }
19. f3 exf3 { exf3 Rxd3 }
20. Rxd3 cxd3 { cxd3 Nxf3 }
21. Kf2 g5 { g5 Kxf3 }
22. Nxf3 e4 { e4 Nxg5 }
23. Bd4+ Kc6 { Kc6 Nxg5 e3+ }
24. Nc7 g4 { g4 a4 }
25. Rf1 a4 { a4 Bxh8 }
26. Rc1+ Kd6 { Kd6 a3 }
27. Ne8+ Ke7 { Ke7 a3 }
28. Rc2 dxc2 { dxc2 Bxh8 }
29. Nd6 exf3 { exf3 Bxh8 }
30. a3 Rh6 { Rh6 Bc3 }
31. Kg3 h4+ { h4+ Kf4 g3 }
32. Kh2 g3+ { g3+ Kg1 }
33. Kxh3 { Kxh3 c1=Q } g2 { g2 Be5 }
34. Bg7 Rh8 { Rh8 Bxh8 Ke6 }
35. Bc3 Rh6 { Rh6 Bb2 }
36. Bg7 Rh8 { Rh8 Bxh8 Ke6 }
37. Bc3 Rh5 { Rh5 Be5 Rxe5 }
38. Bf6+ Nxf6 { Nxf6 Ne4 Ke6 }
39. Nf5+ { Nf5+ Kf7 Kh2 } Ke6 { Ke6 Nxh4 f2 }
40. Ng7+ Ke5 { Ke5 Nxh5 Ng4 }
41. Nf5 Kxf5 { Kxf5 Kh2 h3 }
42. Kh2 c1=R { c1=R Kh3 Rh1#	0-1 }
43. Kh3 Ra7 { Ra7 Kh2 Rh1#	0-1 }
44. Kh2 Rac7 { Rac7 Kh3 Rh1#	0-1 }
45. Kh3 Na6 { Na6 Kh2 Rh1#	0-1 }
46. Kh2 Nb8 { Nb8 Kh3 Rh1#	0-1 }
47. Kh3 Na6 { Na6 Kh2 Rh1#	0-1 }
48. Kh2 Nc5 { Nc5 Kh3 Rh1#	0-1 }
49. Kh3 Rb1 { Rb1 Kh2 Rh1#	0-1 }
50. Kh2 Ra1 { Ra1 Kh3 Rh1#	0-1 }
51. Kh3 Rb1 { Rb1 Kh2 Rh1#	0-1 }
52. Kh2 Ra1 { Ra1 Kh3 Rh1#	0-1 }
53. Kh3 Rc1 { Rc1 Kh2 Rh1#	0-1 }
54. Kh2 Rb1 { Rb1 Kh3 Rh1#	0-1 }
55. Kh3 Ra1 { Ra1 Kh2 Rh1#	0-1 }
56. Kh2 Rb1 { Rb1 Kh3 Rh1#	0-1 }
57. Kh3 Ra1 { Ra1 Kh2 Rh1#	0-1 }
58. Kh2 Rc1 { Rc1 Kh3 Rh1#	0-1 }
59. Kh3 Rd1 { Rd1 Kh2 Rh1#	0-1 }
60. Kh2 Na6 { Na6 Kh3 Rh1#	0-1 }
61. Kh3 Nb4 { Nb4 axb4 Rh1#	0-1 }
62. axb4 a3 { a3 b5 Rh1#	0-1 }
63. Kh2 a2 { a2 b5 Rh1#	0-1 }
64. Kh3 a1=R { a1=R b5 Rh1#	0-1 }
65. Kh2 Ra2 { Ra2 b5 Rh1#	0-1 }
66. Kh3 Raa1 { Raa1 b5 Rh1#	0-1 }
67. Kh2 Ra2 { Ra2 b5 Rh1#	0-1 }
68. Kh3 Ra3 { Ra3 b5 Rh1#	0-1 }
69. Kh2 Raa1 { Raa1 b5 Rh1#	0-1 }
70. Kh3 Ra2 { Ra2 b5 Rh1#	0-1 }
71. Kh2 Raa1 { Raa1 b5 Rh1#	0-1 }
72. Kh3 Ra2 { Ra2 b5 Rh1#	0-1 }
73. Kh2 Ra3 { Ra3 b5 Rh1#	0-1 }
74. Kh3 Ra4 { Ra4 b5 Rh1#	0-1 }
75. Kh2 Ra3 { Ra3 b5 Rh1#	0-1 }
76. Kh3 Ra4 { Ra4 b5 Rh1#	0-1 }
77. Kh2 Ra5 { Ra5 b5 Rh1#	0-1 }
78. bxa5 Rb7 { Rb7 a6 Rh1#	0-1 }
79. Kh3 Ra7 { Ra7 a6 Rh1#	0-1 }
80. a6 Rxa6 { Rxa6 Kh2 Rh1#	0-1 }
81. Kh2 Ra5 { Ra5 Kh3 Rh1#	0-1 }
82. Kh3 Ra4 { Ra4 Kh2 Rh1#	0-1 }
83. Kh2 Ra3 { Ra3 Kh3 Rh1#	0-1 }
84. Kh3 Ra2 { Ra2 Kh2 Rh1#	0-1 }
85. Kh2 Raa1 { Raa1 Kh3 Rh1#	0-1 }
86. Kh3 Ra2 { Ra2 Kh2 Rh1#	0-1 }
87. Kh2 Raa1 { Raa1 Kh3 Rh1#	0-1 }
88. Kh3 Ra3 { Ra3 Kh2 Rh1#	0-1 }
89. Kh2 Ra2 { Ra2 Kh3 Rh1#	0-1 }
90. Kh3 Raa1 { Raa1 Kh2 Rh1#	0-1 }
91. Kh2 Ra2 { Ra2 Kh3 Rh1#	0-1 }
92. Kh3 Raa1 { Raa1 Kh2 Rh1#	0-1 }
93. Kh2 Ra3 { Ra3 Kh3 Rh1#	0-1 }
94. Kh3 Ra4 { Ra4 Kh2 Rh1#	0-1 }
95. Kh2 Ra5 { Ra5 Kh3 Rh1#	0-1 }
96. Kh3 Ra3 { Ra3 Kh2 Rh1#	0-1 }
97. Kh2 Ra4 { Ra4 Kh3 Rh1#	0-1 }
98. Kh3 Ra5 { Ra5 Kh2 Rh1#	0-1 }
99. Kh2 Ra4 { Ra4 Kh3 Rh1#	0-1 }
100. Kh3 Ra5 { Ra5 Kh2 Rh1#	0-1 }
101. Kh2 Ra6 { Ra6 Kh3 Rh1#	0-1 }
102. Kh3 Ra7 { Ra7 Kh2 Rh1#	0-1 }
103. Kh2 Ra6 { Ra6 Kh3 Rh1#	0-1 }
104. Kh3 Ra7 { Ra7 Kh2 Rh1#	0-1 }
105. Kh2 Ra8 { Ra8 Kh3 Rh1#	0-1 }
106. Kh3 Ra6 { Ra6 Kh2 Rh1#	0-1 }
107. Kh2 Ra7 { Ra7 Kh3 Rh1#	0-1 }
108. Kh3 Ra8 { Ra8 Kh2 Rh1#	0-1 }
109. Kh2 Ra7 { Ra7 Kh3 Rh1#	0-1 }
110. Kh3 Ra8 { Ra8 Kh2 Rh1#	0-1 }
111. Kh2 Rb8 { Rb8 Kh3 Rh1#	0-1 }
112. Kh3 Rb7 { Rb7 Kh2 Rh1#	0-1 }
113. Kh2 Rb6 { Rb6 Kh3 Rh1#	0-1 }
114. Kh3 Rb5 { Rb5 Kh2 Rh1#	0-1 }
115. Kh2 Rb4 { Rb4 Kh3 Rh1#	0-1 }
116. Kh3 Rb3 { Rb3 Kh2 Rh1#	0-1 }
117. Kh2 Rb2 { Rb2 Kh3 Rh1#	0-1 }
118. Kh3 Rbb1 { Rbb1 Kh2 Rh1#	0-1 }
119. Kh2 Rb2 { Rb2 Kh3 Rh1#	0-1 }
120. Kh3 Rbb1 { Rbb1 Kh2 Rh1#	0-1 }
121. Kh2 Rb3 { Rb3 Kh3 Rh1#	0-1 }
122. Kh3 Rb2 { Rb2 Kh2 Rh1#	0-1 }
123. Kh2 Rbb1 { Rbb1 Kh3 Rh1#	0-1 }
124. Kh3 Rb2 { Rb2 Kh2 Rh1#	0-1 }
125. Kh2 Rbb1 { Rbb1 Kh3 Rh1#	0-1 }
126. Kh3 Rb3 { Rb3 Kh2 Rh1#	0-1 }
127. Kh2 Rb4 { Rb4 Kh3 Rh1#	0-1 }
128. Kh3 Rb5 { Rb5 Kh2 Rh1#	0-1 }
129. Kh2 Rh1#	0-1
{ Initial time: 30 }
{ Moves to reset clocks: 0 }
{ Time left: White: 16.3977 }
{            Black: 16.4917 }
```

```
ID: 10

Name: Piece Strength Gene
B: -4.28278
N: 4.20362
P: 4.31581
Q: -0.467074
R: 1.97916
INACTIVE

Name: Look Ahead Gene
Mean Game Length: 49.0315
Positions Per Second: 13.9335
INACTIVE

Name: Branch Pruning Gene
Minimum Score Change: 3.91
ACTIVE

Name: Total Force Gene
Scalar: 25.0486
ACTIVE

Name: Freedom to Move Gene
Scalar: 54.1913
INACTIVE

Name: Pawn Advancement Gene
Scalar: -21.6998
INACTIVE

Name: Opponent Pieces Targeted Gene
Scalar: -75.9957
ACTIVE

Name: Sphere of Influence Gene
Legal Bonus: 1.55887
Scalar: 8.41038
ACTIVE

Name: King Confinement Gene
Scalar: 114.206
ACTIVE

Name: King Protection Gene
Scalar: 40.655
ACTIVE

Name: Castling Possible Gene
Scalar: 37.7902
INACTIVE

END
```

```
ID: 51

Name: Piece Strength Gene
B: 1.11332
N: -4.43488
P: 5.73435
Q: -3.43708
R: 1.53077
ACTIVE

Name: Look Ahead Gene
Mean Game Length: 51.7647
Positions Per Second: 52.405
ACTIVE

Name: Branch Pruning Gene
Minimum Score Change: -8.23496
ACTIVE

Name: Total Force Gene
Scalar: 26.6085
ACTIVE

Name: Freedom to Move Gene
Scalar: 2.21285
INACTIVE

Name: Pawn Advancement Gene
Scalar: 39.7031
ACTIVE

Name: Opponent Pieces Targeted Gene
Scalar: -21.1828
INACTIVE

Name: Sphere of Influence Gene
Legal Bonus: 3.73808
Scalar: -35.3508
INACTIVE

Name: King Confinement Gene
Scalar: -60.9795
INACTIVE

Name: King Protection Gene
Scalar: -3.68461
ACTIVE

Name: Castling Possible Gene
Scalar: 38.0428
INACTIVE

END
```

**Comment by MarkZH** [comment]
First hypothesis:

The problem occurs when there is more than one path to checkmate. If the longer path is found first, the search cuts off due to alpha-beta pruning since wins are given a score of +infinity. First change is to make wins worth `std::numeric_limits<double>::max()` to keep the score finite and within the initial alpha and beta limits to allow for shorter paths to checkmate to be found.

**Comment by MarkZH** [comment]
Above fix seems to have fixed the major problem.

Now, when a Genetic_AI tries to distance itself from losing, it seems to return a line that misses the vulnerability to a nearer checkmate. See below for an example:

```
[White "Genetic AI 668"]
[Black "Genetic AI 576"]
[Round "811"]
[Result "1-0"]

1. d4 { d4 d5 } d5 { d5 e4 }
2. a4 a5
3. g4 { g4 Bxg4 } Bxg4 { Bxg4 e4 }
4. Na3 b5
5. Nxb5 f5 { f5 e4 }
6. b4 axb4
7. a5 g5 { g5 e4 }
8. a6 Nxa6 { Nxa6 e4 }
9. Bxg5 { Bxg5 e5 } Nc5
10. Rxa8 Nd3+
11. cxd3 { cxd3 Qxa8 } Qxa8 { Qxa8 f4 }
12. Nxc7+ Kd7 { Kd7 Qa4+ }
13. Nxa8 e5 { e5 Qa4+ }
14. Qa4+ { Qa4+ Kc8 } Kc8 { Kc8 dxe5 }
15. Qxb4 Bxb4+ { Bxb4+ Bd2 }
16. Bd2 { Bd2 Bxd2+ } Bxd2+ { Bxd2+ Kxd2 }
17. Kxd2 { Kxd2 exd4 f4 } exd4 { exd4 e4 }
18. f4 { f4 Bxe2 } h5 { h5 e4 }
19. h4 { h4 Kb7 } Rh6 { Rh6 Nc7 }
20. Nb6+ { Nb6+ Kb8 } Rxb6 { Rxb6 Kc2 Kc7 }
21. Kc1 { Kc1 Bxe2 } Kc7 { Kc7 e4 }
22. Nh3 { Nh3 Rb1+ } Ra6
23. Kb1 { Kb1 Bxe2 } Ra5
24. Bg2 { Bg2 Ra1+ } Rb5+ { Rb5+ Kc2 }
25. Kc1 { Kc1 Rb1+ } Rc5+ { Rc5+ Kb1 }
26. Kb2 { Kb2 Rc2+ } Rb5+ { Rb5+ Ka3 }
27. Ka3 { Ka3 Ra5+ } Ra5+ { Ra5+ Kb4 }
28. Kb3 { Kb3 Ra3+ } Rb5+ { Rb5+ Ka3 }
29. Ka4 { Ka4 Ra5+ } Bxe2 { Bxe2 Kxb5 }
30. Kxb5 { Kxb5 Bxd3+ } Bxd3+ { Bxd3+ Kb4 }
31. Kc5 { Kc5 Bc4 Bxd5 } Nh6 { Nh6 Kxd4 }
32. Bxd5 { Bxd5 Ng4 } Ng4 { Ng4 Kxd4 }
33. Kxd4 { Kxd4 Bf1 } Nf2 { Nf2 Bc6 }
34. Nxf2 { Nxf2 Bf1 } Bc2 { Bc2 Bc6 }
35. Kc5 { Kc5 Bd1 } Kd7 { Kd7 Nd1 }
36. Rh3 { Rh3 Kd8 Bc6 } Ke7 { Ke7 Be6 }
37. Rh1 { Rh1 Kd7 Ne4 } Kd7 { Kd7 Nd1 }
38. Rh3 { Rh3 Kd8 Bc6 } Ke7 { Ke7 Be6 }
39. Re3+ { Re3+ Be4 } Kf6 { Kf6 Re6+ }
40. Nh3 { Nh3 Ba4 Re6+ } Kg7 { Kg7 Re5 Be4 }
41. Re7+ { Re7+ Kf6 } Kf6 { Kf6 Rf7+ Kg6 }
42. Rf7+ { Rf7+ Kg6 Rxf5 } Kg6
43. Ng1 { Ng1 Bb1 } Ba4 { Ba4 Be6 }
44. Nh3 { Nh3 Kh6 Rxf5 } Be8 { Be8 Bg2 }
45. Ra7 { Ra7 Ba4 } Kf6 { Kf6 Rf7+ }
46. Ng5 { Ng5 Kg6 Bf3 } Bg6 { Bg6 Be6 Be8 }
47. Nh7+ { Nh7+ Bxh7 Rxh7 Kg6 } Bxh7
48. Rxh7 { Rxh7 Kg6 } Kg6
49. Rf7 { Rf7 Kh6 Rxf5 } Kh6
50. Be6 { Be6 Kg6 Bxf5+ } Kg6
51. Rxf5 { Rxf5 Kh6 Rxh5+ } Kh7 { Kh7 Rxh5+ Kg7 }
52. Rxh5+ { Rxh5+ Kg7 f5 } Kg7 { Kg7 Rh6 Kxh6 h5 }
53. f5 { f5 Kf6 Rh8 } Kf6 { Kf6 Kd6 Kg7 f6+ }
54. Rh8 { Rh8 Ke5 f6 } Ke5 { Ke5 h5 Ke4 }
55. Ra8 { Ra8 Kf6 Ra2 } Ke4 { Ke4 f6 Kd3 }
56. f6 { f6 Kd3 Ra2 } Ke5 { Ke5 f7 Kxe6 h5 }
57. f7 { f7 Kxe6 h5 } Kxe6 { Kxe6 Ra7 Ke5 }
58. f8=Q { f8=Q Ke5 Re8#	1-0 } Kd7 { Kd7 Re8 Kc7 Re7#	1-0 }
59. Qd6#	1-0
{ Initial time: 30 }
{ Moves to reset clocks: 0 }
{ Time left: White: 4.67683 }
{            Black: 8.20148 }
```
Move 58 for Black seems too optimistic. It may actually be that the alpha-beta pruning is working and there really is nothing better. Every move is a loss and any move is returned. It may just be the commentary that is misleading due to the early break.

**Comment by MarkZH** [comment]
Instead of alpha and beta just being numbers, they should be Game_Tree_Node_Results and compared the same way. Create

`bool operator>(const Game_Tree_Node_Result&, const Game_Tree_Node_Result&)`

`bool operator==(const Game_Tree_Node_Result&, const Game_Tree_Node_Result&)`

`bool operator>=(const Game_Tree_Node_Result&, const Game_Tree_Node_Result&)`

As well as corresponding +/-infinte values for `Game_Tree_Node_Result`. The inequality comparisons should act the same as the current `game_tree_search()` comparison:

1. Prefer the higher score.
2. Else, prefer a win at shallower depths.
3. Else, prefer a loss at greater depths.

**Comment by MarkZH** [comment]
Above comment implemented in commit 22fdb0022aa70e4eea0ad69b1e7f99637bc7327f.

**Comment by MarkZH** [comment]
Logic errors discovered without this capability. Analyzing game records and commentary is enough. The Game_Tree_Node_Result comparisons fixed the original problem in commit 22fdb0022aa70e4eea0ad69b1e7f99637bc7327f.
